pkg/controller/controller_instance.go

```diff
@@ -1457,35 +1466,52 @@ func (c *controller) processProvisionSuccess(instance *v1beta1.ServiceInstance,
 	return nil
 }
 
-// processProvisionFailure handles the logging and updating of a
+// processTerminalProvisionFailure handles the logging and updating of a
 // ServiceInstance that hit a terminal failure during provision reconciliation.
-func (c *controller) processProvisionFailure(instance *v1beta1.ServiceInstance, readyCond, failedCond *v1beta1.ServiceInstanceCondition, shouldMitigateOrphan bool) error {
-	currentReconciledGeneration := instance.Status.ReconciledGeneration
+func (c *controller) processTerminalProvisionFailure(instance *v1beta1.ServiceInstance, readyCond, failedCond *v1beta1.ServiceInstanceCondition, shouldMitigateOrphan bool) error {
 	if failedCond == nil {
 		return fmt.Errorf("failedCond must not be nil")
 	}
+	return c.processProvisionFailure(instance, readyCond, failedCond, shouldMitigateOrphan)
+}
+
+// processTemporaryProvisionFailure handles the logging and updating of a
+// ServiceInstance that hit a temporary error during provision reconciliation.
+func (c *controller) processTemporaryProvisionFailure(instance *v1beta1.ServiceInstance, readyCond *v1beta1.ServiceInstanceCondition, shouldMitigateOrphan bool) error {
+	return c.processProvisionFailure(instance, readyCond, nil, shouldMitigateOrphan)
+}
+
+// processProvisionFailure handles the logging and updating of a
+// ServiceInstance that hit a temporary or a terminal failure during provision
+// reconciliation.
+func (c *controller) processProvisionFailure(instance *v1beta1.ServiceInstance, readyCond, failedCond *v1beta1.ServiceInstanceCondition, shouldMitigateOrphan bool) error {
+	currentReconciledGeneration := instance.Status.ReconciledGeneration
 
 	if readyCond != nil {
 		c.recorder.Event(instance, corev1.EventTypeWarning, readyCond.Reason, readyCond.Message)
 		setServiceInstanceCondition(instance, v1beta1.ServiceInstanceConditionReady, readyCond.Status, readyCond.Reason, readyCond.Message)
 	}
 
-	c.recorder.Event(instance, corev1.EventTypeWarning, failedCond.Reason, failedCond.Message)
-	setServiceInstanceCondition(instance, v1beta1.ServiceInstanceConditionFailed, failedCond.Status, failedCond.Reason, failedCond.Message)
+	var errorMessage error
+	if failedCond != nil {
+		c.recorder.Event(instance, corev1.EventTypeWarning, failedCond.Reason, failedCond.Message)
+		setServiceInstanceCondition(instance, v1beta1.ServiceInstanceConditionFailed, failedCond.Status, failedCond.Reason, failedCond.Message)
+		errorMessage = fmt.Errorf(failedCond.Message)
+	} else {
+		resetServiceInstanceCondition(instance, v1beta1.ServiceInstanceConditionFailed)
+		errorMessage = fmt.Errorf(readyCond.Message)
+	}
 
-	// Need to vary return error depending on whether the worker should
-	// requeue this resource.
-	var err error
 	if shouldMitigateOrphan {
+		// TODO nilebox: if failedCond == nil, we lose the original error reason/message
+		// in the status completely there. Should we keep the original reason/message instead?
```

Comment:
I propose that we add a new OrphanMitigation condition that is true when the instance is undergoing orphan mitigation. The Ready condition is not changed when starting and stopping orphan mitigation. Presumably, the Ready condition will already be false when orphan mitigation starts. Although I'd also be fine with relying solely on `OrphanMitigiationInProgress` instead of adding a new condition. Either way, I suggest that we stop changing the Ready condition with the starting orphan mitigation reason. Of course, this is all follow-on work.

Reply:
@staebler I personally don't have objection against OrphanMitigation condition, we can discuss this separately with @pmorie and others.
As you suggest (and as my TODO suggests too), I will keep the original reason/message for now, and probably leave a TODO that it would be nice to somehow reflect the started orphan mitigation in the status somehow.

Reply:
Raised an issue #1771 to discuss this.